game-core/src/main/java/games/strategy/engine/framework/startup/ui/posted/game/pbem/EmailSenderEditorViewModel.java

```diff
@@ -0,0 +1,194 @@
+    return smtpServer != null
+        && !smtpServer.isBlank()
+        && !toAddress.isBlank()
+        && PlayerEmailValidation.isValid(toAddress)
+        && StringUtils.isPositiveInt(smtpPort)
+        && ClientSetting.emailUsername.isSet()
+        && ClientSetting.emailUsername.getValueOrThrow().length > 0
+        && ClientSetting.emailPassword.isSet()
+        && ClientSetting.emailPassword.getValueOrThrow().length > 0;
+  }
+
+  void setSelectedProvider(final String provider) {
+    Preconditions.checkNotNull(provider);
+    if (!selectedProvider.equals(provider)) {
+      selectedProvider = provider;
+      ClientSetting.emailProvider.setValueAndFlush(selectedProvider);
+
+      EmailProviderPreset.lookupByName(provider)
+          .ifPresentOrElse(
+              preset -> {
+                smtpServer = preset.getServer();
+                smtpPort = String.valueOf(preset.getPort());
+                useTls = preset.isUseTlsByDefault();
+
+                ClientSetting.emailServerHost.setValue(smtpServer);
+                ClientSetting.emailServerPort.setValue(preset.getPort());
+                ClientSetting.emailServerSecurity.setValueAndFlush(useTls);
+              },
+              () -> {
+                smtpServer = "";
+                smtpPort = "";
+                useTls = true;
+
+                ClientSetting.emailServerHost.resetValue();
+                ClientSetting.emailServerPort.resetValue();
+                ClientSetting.emailServerSecurity.setValueAndFlush(useTls);
+              });
+
+      invokeCallbacks();
+    }
+  }
+
+  private void invokeCallbacks() {
+    view.viewModelChanged(this);
+    Optional.ofNullable(validatedFieldsChangedListener).ifPresent(Runnable::run);
+  }
+
+  boolean showServerOptions() {
+    return selectedProvider.equals(GENERIC_SMTP);
+  }
+
+  boolean showEmailOptions() {
+    return !isEmailProviderDisabled();
+  }
+
+  private boolean isEmailProviderDisabled() {
+    return selectedProvider.equals(PROVIDER_DISABLED);
+  }
+
+  void setSmtpServer(final String smtpServer) {
+    Preconditions.checkNotNull(smtpServer);
+    if (!this.smtpServer.equals(smtpServer)) {
+      this.smtpServer = smtpServer.trim();
+      ClientSetting.emailServerHost.setValueAndFlush(smtpServer);
+      invokeCallbacks();
+    }
+  }
+
+  boolean isSmtpServerValid() {
+    return isEmailProviderDisabled() || !smtpServer.isBlank();
+  }
+
+  void setSmtpPort(final String smtpPort) {
+    Preconditions.checkNotNull(smtpPort);
+    if (!this.smtpPort.equals(smtpPort)) {
+      this.smtpPort = smtpPort;
+      if (StringUtils.isPositiveInt(smtpPort)) {
+        ClientSetting.emailServerPort.setValueAndFlush(Integer.valueOf(smtpPort.trim()));
+      } else {
+        ClientSetting.emailServerPort.resetValue();
+        ClientSetting.flush();
+      }
+      invokeCallbacks();
+    }
+  }
+
+  boolean isSmtpPortValid() {
+    return isEmailProviderDisabled() || StringUtils.isPositiveInt(smtpPort);
```

Comment:
If you want to be precise, you'd have to check if the port is inside `[1, 65535]`

Reply:
Indeed, that was a consideration, still is I suppose. Though, I'm not entirely sure if we really need to go to that length of validation. Just notably it makes potential error messaging perhaps overly complex for a user. If they have a port number that looks right, but is just wrong, they'll have trouble connecting either way.